searx/shared/__init__.py

```diff
@@ -0,0 +1,31 @@
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+import logging
+
+logger = logging.getLogger('searx.shared')
+
+try:
+    import uwsgi
+except:
+    # no uwsgi
+    from .shared_simple import SimpleSharedDict as SharedDict, schedule
+    logger.info('Use shared_simple implementation')
+else:
+    try:
+        uwsgi.cache_update('dummy', b'dummy')
+        if uwsgi.cache_get('dummy') != b'dummy':
+            raise Exception()
+    except:
+        # uwsgi.ini configuration problem: disable all scheduling
+        logger.error('uwsgi.ini configuration error, add this line to your uwsgi.ini\n'
+                     'cache2 = name=searxcache,items=2000,blocks=2000,blocksize=4096,bitmap=1')
+        from .shared_simple import SimpleSharedDict as SharedDict
+
+        def schedule(delay, func, *args):
+            pass
+    else:
+        # uwsgi
+        from .shared_uwsgi import UwsgiCacheSharedDict as SharedDict, schedule
+        logger.info('Use shared_uwsgi implementation')
+
+storage = SharedDict()
```

Comment:
Why is it a dict? What else are you planning to store in `storage` that you need to have in the same key-value storage?

Reply:
* `ShareDict` encoding / decoding is simple : uwsgi deals with `bytes` not a Python object.
* the caller declares what kind of data is expected (not Python spirit I know, but here it is about cooperation between multiple process, I prefer something more declarative)

The alternative is to serialize the Python object using `pickle` or something equivalent, so `SharedDict`, so we can have `storage[key] = value` and `storage[key]`.

> What else are you planning to store in storage that you need to have in the same key-value storage?

Response times: https://github.com/searx/searx/pull/477 : here pickling each number is not very efficient.

Important note: the uwsgi implementation won't work with https://github.com/searx/searx/pull/2313

An alternative implementation can use [anonymous memory map](https://docs.python.org/3/library/mmap.html) which also provides `bytes` (without a hash table).